### PR TITLE
bpo-40408: Fix support of nested type variables in GenericAlias.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue40408

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
@gvanrossum please take a look at this.

File: Objects/genericaliasobject.c [hunk 1]
```diff
@@ -182,28 +182,57 @@ tuple_index(PyObject *self, Py_ssize_t len, PyObject *item)
...
-                Py_INCREF(t);
-                PyTuple_SET_ITEM(parameters, iparam, t);
-                iparam++;
+            iparam += tuple_add(parameters, iparam, t);
+        }
+        else {
+            _Py_IDENTIFIER(__parameters__);
+            PyObject *subparams;
+            if (_PyObject_LookupAttrId(t, &PyId___parameters__, &subparams) < 0) {
+                Py_DECREF(parameters);
```
Comment (serhiy-storchaka) [inline Objects/genericaliasobject.c:219]: I use ducktyping here to support both `GenericAlias` and `_GenericAlias` (`list[list[T]]` and `list[List[T]]`). There is some difference with `_GenericAlias`, but this behavior is not specified and not tested.

File: Lib/test/test_genericalias.py [hunk 2]
```diff
@@ -170,10 +172,7 @@ def test_exposed_type(self):
         self.assertEqual(a.__parameters__, ())
 
     def test_parameters(self):
-        from typing import TypeVar
-        T = TypeVar('T')
-        K = TypeVar('K')
-        V = TypeVar('V')
+        from typing import List, Dict
```
Comment (gvanrossum) [inline Lib/test/test_genericalias.py]: Why not move these imports to the toplevel too? I don't think the local import served a purpose before either, and they were done inconsistently.

File: Lib/test/test_genericalias.py [hunk 3]
```diff
@@ -195,14 +194,34 @@ def test_parameters(self):
         L1 = list[T]
         self.assertEqual(L1.__args__, (T,))
         self.assertEqual(L1.__parameters__, (T,))
+        L2 = list[List[T]]
+        self.assertEqual(L2.__args__, (List[T],))
```
Comment (gvanrossum) [inline Lib/test/test_genericalias.py]: Probably should separately also test for `list[list[T]]` etc. (In `test_parameter_chaining()` below you do do test both scenarios separately.)

File: Objects/genericaliasobject.c (hunk 1 above)
Comment (gvanrossum) [reply Objects/genericaliasobject.c:219]: Indeed, we intentionally aligned the attribute names (`__origin__`, `__attrs__`, `__parameters__`) to enable this.

File: Objects/genericaliasobject.c [hunk 4]
```diff
@@ -182,28 +182,57 @@ tuple_index(PyObject *self, Py_ssize_t len, PyObject *item)
...
+                return NULL;
             }
+            if (subparams && PyTuple_Check(subparams)) {
+                Py_ssize_t len2 = PyTuple_GET_SIZE(subparams);
+                len += len2 - 1 - (iarg - iparam);
+                if (_PyTuple_Resize(&parameters, len) < 0) {
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: If len2 is zero you can skip the whole thing. Also, why use such a complex formula for the new length? At this point `len` can be an over- or an under-estimation, and at the end we right-size it, so we might as well do something simpler here like `len += len2`.

File: Objects/genericaliasobject.c [hunk 5]
```diff
@@ -254,15 +285,46 @@ ga_getitem(PyObject *self, PyObject *item)
...
+        }
+        else {
+            _Py_IDENTIFIER(__parameters__);
+            PyObject *subparams;
+            if (_PyObject_LookupAttrId(arg, &PyId___parameters__, &subparams) < 0) {
+                Py_DECREF(newargs);
+                return NULL;
+            }
+            if (subparams && PyTuple_Check(subparams)) {
+                Py_ssize_t nsubargs = PyTuple_GET_SIZE(subparams);
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: Again, you can reuse `arg` if `nsubargs` is zero.

File: Objects/genericaliasobject.c [hunk 6]
```diff
@@ -254,15 +285,46 @@ ga_getitem(PyObject *self, PyObject *item)
         if (typevar) {
             Py_ssize_t iparam = tuple_index(alias->parameters, nparams, arg);
             assert(iparam >= 0);
-            if (is_tuple) {
-                arg = PyTuple_GET_ITEM(item, iparam);
+            arg = args[iparam];
+            Py_INCREF(arg);
+        }
+        else {
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c:342]: Can you turn this entire `else` block into a helper function? (Assuming there aren't too many variables to pass in.) Or at least a large portion of it. The top-level logic is hard to follow (I almost needed to resort to pen and paper to convince myself it is correct :-).

Also we could use a few examples, e.g.
```
t = list[T];          t[int]      -> newargs = [int]
t = dict[str, T];     t[int]      -> newargs = [str, int]
t = dict[T, list[S]]; t[str, int] -> newargs = [str, list[int]]
```

File: Objects/genericaliasobject.c [hunk 7]
```diff
@@ -234,6 +263,7 @@ ga_getitem(PyObject *self, PyObject *item)
     }
     int is_tuple = PyTuple_Check(item);
     Py_ssize_t nitem = is_tuple ? PyTuple_GET_SIZE(item) : 1;
+    PyObject **args = is_tuple ? &PyTuple_GET_ITEM(item, 0) : &item;
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: Maybe rename this to `items`? Or `argitems`? I was quite confused for a while by the distinction between `alias->args` (the `__args__` attribute) vs. `args` (the arguments to `__getitem__`).

File: Objects/genericaliasobject.c [hunk 8]
```diff
@@ -254,15 +285,46 @@ ga_getitem(PyObject *self, PyObject *item)
...
+            Py_INCREF(arg);
+        }
+        else {
+            _Py_IDENTIFIER(__parameters__);
+            PyObject *subparams;
+            if (_PyObject_LookupAttrId(arg, &PyId___parameters__, &subparams) < 0) {
+                Py_DECREF(newargs);
+                return NULL;
+            }
+            if (subparams && PyTuple_Check(subparams)) {
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: When could `subparams` be zero here? You already check for failure from the lookup. (Same above in make_parameters() I guess.)

**Review by gvanrossum — COMMENTED** [review]
LG, just style suggestions.

File: Lib/test/test_genericalias.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_genericalias.py]: I did it initially. But in other test we import `Callable` which hides the `Callable` imported from `collections.abc`. To avoid this and possible future conflicts (with `Sequence`, `Iterable`, etc) I prefer to import they locally. It is easy to forgot which names in `typing` and `collections.abc` conflict and which do not.

Added a test for `Callable`.

File: Objects/genericaliasobject.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c]: The purpose is to avoid resizing the tuple if we have enough space reserved. Although I need to add an additional check here.

1 place was already reserved for the arguments itself, so we should add maximum `len2 - 1` to the length. But since some arguments could be skipped before we have `iarg - iparam` additional free places.

File: Objects/genericaliasobject.c (hunk 6 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c:342]: Moved.

I am not good in commenting, and in this particular case it is especially difficult to me to write something meaningful. Could you please help me and propose concrete text and where it should be inserted? I am sorry, but it is just not my skill.

File: Objects/genericaliasobject.c (hunk 8 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c]: It could be zero when `arg` does not have the `__parameters__` attribute.

There are three possibility:
* There is the attribute. `_PyObject_LookupAttrId` returns `1` and sets the output pointer to non-NULL.
* There is no the attribute. `_PyObject_LookupAttrId` returns `0` and sets the output pointer to NULL. No exception is set (if AttributeError was raised, it will be silenced).
* Getting the attribute raises an error. `_PyObject_LookupAttrId` returns `-1`, sets the output pointer to NULL and sets an exception.

It looks a little complicated, but allows to write the most simple code and have the best performance (`AttributeError` is not even raised just to be silenced, and you avoid to call `PyErr_Occurred()` in performance critical code, and explicitly check `PyErr_ExceptionMatches(PyExc_AttributeError)`).

`_PyObject_LookupAttrId` was initially added for performance, but later it replaced `PyObject_GetAttr` in many sites, because it allowed easier to write the correct code (when you need to ignore AttributeError but not other exceptions).

File: Objects/genericaliasobject.c (hunk 5 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c]: Good catch. There was a bug (`dict[T, list[int]][str]` did not work), fixed and added a test.

File: Objects/genericaliasobject.c (hunk 7 above)
Comment (serhiy-storchaka) [reply Objects/genericaliasobject.c]: Done.

File: Lib/test/test_genericalias.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/test_genericalias.py]: Done.

File: Objects/genericaliasobject.c (hunk 6 above)
Comment (gvanrossum) [reply Objects/genericaliasobject.c:342]: You're underestimating yourself. (Your English has vastly improved over the years -- at this point I think it's better than Victor's. :-) But I will suggest what I think is still needed.

File: Objects/genericaliasobject.c [hunk 9]
```diff
@@ -263,13 +308,19 @@ ga_getitem(PyObject *self, PyObject *item)
     }
     int is_tuple = PyTuple_Check(item);
     Py_ssize_t nitem = is_tuple ? PyTuple_GET_SIZE(item) : 1;
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: I'd rename `nitem` to `nitems`.

File: Objects/genericaliasobject.c [hunk 10]
```diff
@@ -244,6 +247,48 @@ make_parameters(PyObject *args)
     return parameters;
 }
 
+/* If obj is a generic alias, substitute type variables params
+   with substitutions argitems. */
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]:
```suggestion
/* If obj is a generic alias, substitute type variables params
   with substitutions argitems.  For example, if obj is list[T],
   params is (T, S), and argitems is (str, int), return list[str].
   If obj doesn't have a __parameters__ attribute or that's not
   a non-empty tuple, return obj, INCREF'ed. */
```

File: Objects/genericaliasobject.c [hunk 11]
```diff
@@ -244,6 +247,48 @@ make_parameters(PyObject *args)
...
+            for (Py_ssize_t i = 0; i < nsubargs; ++i) {
+                PyObject *arg = PyTuple_GET_ITEM(subparams, i);
+                Py_ssize_t iparam = tuple_index(params, nparams, arg);
+                if (iparam >= 0) {
+                    arg = argitems[iparam];
+                }
+                Py_INCREF(arg);
+                PyTuple_SET_ITEM(subargs, i, arg);
+            }
+            obj = PyObject_GetItem(obj, subargs);
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: Since this is the operation we've been preparing for, I'd set it apart with a blank line before and after.

File: Objects/genericaliasobject.c [hunk 12]
```diff
@@ -244,6 +247,48 @@ make_parameters(PyObject *args)
...
+static PyObject *
+subs_tvars(PyObject *obj, PyObject *params, PyObject **argitems)
+{
+    _Py_IDENTIFIER(__parameters__);
+    PyObject *subparams;
+    if (_PyObject_LookupAttrId(obj, &PyId___parameters__, &subparams) < 0) {
+        return NULL;
+    }
+    if (subparams && PyTuple_Check(subparams)) {
+        Py_ssize_t nparams = PyTuple_GET_SIZE(params);
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c:264]: This could be moved into the `if (nsubargs != 0) {` block below.

(Also the logic would be slightly simpler if the check for subparams being non-empty was part of the `if (subparams && PyTuple_Check(subparams)) {` check above.)

File: Objects/genericaliasobject.c [hunk 13]
```diff
@@ -285,45 +336,15 @@ ga_getitem(PyObject *self, PyObject *item)
...
-                    return NULL;
-                }
-            }
-            else {
-                Py_INCREF(arg);
-            }
-            Py_XDECREF(subparams);
         }
         PyTuple_SET_ITEM(newargs, iarg, arg);
     }
```
Comment (gvanrossum) [inline Objects/genericaliasobject.c:350]: Comment for the next line: (again) since it is the main event we've been working up to, offset it with a blank line before and after.

**Review by gvanrossum — COMMENTED** [review]
Some more suggestions. Thanks for explaining LookupAttrId, I had forgotten its API. :-)

**Review by gvanrossum — APPROVED** [review]
Perfect!

**Comment by serhiy-storchaka** [comment]
Thank you very much for your review!
